File: radial_distortion_correction.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <cmath>

using namespace cv;
using namespace std;

// Global variables to store points and image
vector<Point2d> points;
Mat image;

// Mouse callback function to capture points
void click_event(int event, int x, int y, int, void*) {
    if (event == EVENT_LBUTTONDOWN) {
        points.push_back(Point2d(x, y));
        if (points.size() <= 3) {
            circle(image, Point(x, y), 5, Scalar(0, 255, 0), -1);
            imshow("Image", image);
        }
    }
}

// Function to apply the radial distortion
Point2d apply_radial_distortion(Point2d point, Point2d center, double w) {
    if (w == 0) return point;
    double ru = norm(point - center);
    double rd = atan(2 * ru * tan(w / 2)) / w;
    Point2d distorted = (point - center)* rd / ru + center;
    return distorted;
}

// Function to correct the radial distortion
Point2d correct_radial_distortion(Point2d point, Point2d center, double w) {
    if (w == 0) return point;
    double rd = norm(point - center);
    double ru = tan(rd * w) / (2 * tan(w / 2));
    Point2d corrected = (point - center) * ru / rd + center;
    return corrected;
}

// Objective function to minimize
double objective_function(const double w, const vector<Point2d>& points, Point2d center) {
    Point2d p1 = correct_radial_distortion(points[0], center, w);
    Point2d p2 = correct_radial_distortion(points[1], center, w);
    Point2d p3 = correct_radial_distortion(points[2], center, w);

    Point2d v1 = (p2 - p1) / norm(p2 - p1);
    Point2d v2 = (p3 - p1) / norm(p3 - p1);
    double loss = norm(v1.cross(v2));
    return loss;
}

// Gradient Descent implementation
double gradient_descent(const vector<Point2d>& points, Point2d center, double learning_rate = 1e-5, double tolerance = 1e-12, int max_iter = 100) {
    double w = 0;
    double eps = 1e-6;
    for (int iter = 0; iter < max_iter; iter++) {
        double w_eps = w + eps;
        double grad = (objective_function(w_eps, points, center) - objective_function(w, points, center)) / eps;
        double new_w = w - learning_rate * grad;

        double delta = fabs(new_w - w);
        if (delta < tolerance) {
            break;
        }

        w = new_w;
        double loss = objective_function(w, points, center);
        printf("[%d] loss = %lf, w = %lf\n", iter, loss, w);
    }

    return w;
}

// Main function
int main() {
    // Read the image
    image = imread("sample_radial.png");
    if (image.empty()) {
        cout << "Image not found!" << endl;
        return -1;
    }

    namedWindow("Image", WINDOW_AUTOSIZE);
    setMouseCallback("Image", click_event);

    cout << "Click on three points in the image" << endl;
    imshow("Image", image);
    waitKey(0);

    if (points.size() < 3) {
        cout << "Not enough points selected!" << endl;
        return -1;
    }

    // Perform gradient descent to find k1 and k2
    double height = image.rows;
    double width = image.cols;
    Point2d center(width / 2, height / 2);
    double w = gradient_descent(points, center);
    cout << "Found radial distortion coefficients: w = " << w << endl;

    // Apply distortion correction to the image
    Mat new_image = Mat::zeros(image.size(), image.type());
    for (int y = 0; y < height; y++) {
        for (int x = 0; x < width; x++) {
            Point2d distorted_point = apply_radial_distortion(Point2d(x, y), center, w);
            int xd = cvRound(distorted_point.x);
            int yd = cvRound(distorted_point.y);
            if (0 <= xd && xd < width && 0 <= yd && yd < height) {
                new_image.at<Vec3b>(y, x) = image.at<Vec3b>(yd, xd);
            }
        }
    }

    imshow("Corrected Image", new_image);
    waitKey(0);
    destroyAllWindows();

    return 0;
}

File: radial_distortion_correction.py
import numpy as np
import math
import matplotlib.pyplot as plt
from tkinter import Tk, Canvas, Button, filedialog
plt.rcParams.update({'font.size': 14})

# Global variables
points = []
image = None
canvas_img = None
canvas_widget = None

# Load an image using a file dialog
def load_image():
    global image, canvas_img, canvas_widget
    file_path = filedialog.askopenfilename()
    if not file_path:
        return
    img = plt.imread(file_path)
    if img is not None:
        image = img
        canvas_img = np.copy(image)
        display_image_on_canvas()

# Display the image on the Tkinter canvas
def display_image_on_canvas():
    global canvas_widget, canvas_img
    if canvas_widget is not None:
        canvas_widget.delete("all")  # Clear existing canvas content
    height, width = canvas_img.shape[:2]
    # Resize the canvas to match the image size
    canvas.config(width=width, height=height)
    # Convert image array to PhotoImage-compatible format
    from PIL import Image, ImageTk
    img_pil = Image.fromarray((canvas_img * 255).astype(np.uint8))  # Ensure values are 0-255
    img_tk = ImageTk.PhotoImage(img_pil)
    canvas_widget = canvas.create_image(0, 0, anchor="nw", image=img_tk)
    canvas.image_tk = img_tk  # Keep a reference to prevent garbage collection

# Mouse callback function to capture points
def click_event(event):
    global points
    if len(points) < 3:
        x, y = event.x, event.y
        points.append((x, y))
        canvas.create_oval(x - 5, y - 5, x + 5, y + 5, fill="red")
        if len(points) == 3:
            process_points()

# Function to apply the radial distortion
def apply_radial_distortion(point, center, w):
    if w == 0 or point == center:
        return point
    ru = np.linalg.norm(np.array(point) - np.array(center))
    rd = np.arctan(2 * ru * np.tan(w / 2)) / w
    distorted = (np.array(point) - np.array(center)) * rd / ru + np.array(center)
    return distorted

# Function to correct the radial distortion
def correct_radial_distortion(point, center, w):
    if w == 0 or point == center:
        return point
    rd = np.linalg.norm(np.array(point) - np.array(center))
    ru = np.tan(rd * w) / (2 * np.tan(w / 2))
    corrected = (np.array(point) - np.array(center)) * ru / rd + np.array(center)
    return corrected
    
# Objective function to minimize
def objective_function(w, points, center):
    p1 = correct_radial_distortion(points[0], center, w)
    p2 = correct_radial_distortion(points[1], center, w)
    p3 = correct_radial_distortion(points[2], center, w)

    v1 = (np.array(p2) - np.array(p1)) / np.linalg.norm(np.array(p2) - np.array(p1))
    v2 = (np.array(p3) - np.array(p1)) / np.linalg.norm(np.array(p3) - np.array(p1))
    loss = np.linalg.norm(np.cross(v1, v2))
    return loss
    
# Perform gradient descent to find optimal w
def gradient_descent(points, center, learning_rate=1e-6, tolerance=1e-12, max_iter=1000):
    w = 0
    eps = 1e-7
    for iter in range(max_iter):
        w_eps = w + eps
        grad = (objective_function(w_eps, points, center) - objective_function(w, points, center)) / eps
        new_w = w - learning_rate * grad

        delta = abs(new_w - w)
        if delta < tolerance:
            break

        w = new_w
        loss = objective_function(w, points, center)
        print(f"[{iter}] loss = {loss}, w = {w}")
    return w

# Process the selected points and show corrected image
def process_points():
    global points, image
    if len(points) < 3:
        print("Not enough points selected!")
        return

    height, width = image.shape[:2]
    center = (width / 2, height / 2)

    # Perform gradient descent
    w = gradient_descent(points, center)
    print(f"Found radial distortion coefficients: w = {w}")

    # Apply distortion correction to the image
    new_image = np.zeros_like(image)
    for y in range(height):
        for x in range(width):
            distorted_point = apply_radial_distortion((x, y), center, w)
            xd = int(round(distorted_point[0]))
            yd = int(round(distorted_point[1]))
            if 0 <= xd < width and 0 <= yd < height:
                new_image[y, x] = image[yd, xd]

    # Show corrected image
    plt.imshow(new_image)
    plt.title("Corrected Image")
    plt.axis('off')
    plt.show()

# Initialize Tkinter
root = Tk()
root.title("Radial Distortion Correction")

# Create canvas for image display
canvas = Canvas(root)
canvas.pack()

# Bind mouse click events to the canvas
canvas.bind("<Button-1>", click_event)

# Add buttons
load_btn = Button(root, text="Load Image", command=load_image)
load_btn.pack()

root.mainloop()